### ISSUE TITLE
Persist queued queries 

### ISSUE THREAD

**Issue by MrLoh** [helfer/apollo-link-queue#3]
Hi this looks very promising. However to create a full offline support what is missing is a way to persist the queued queries, so when a user is for example on holiday and doesn’t have internet, the local state of the app Including his mutations will be save when the app is restarted. Would you have any idea how to persist the queued queries, could the operation simply be serialized?

**Comment by helfer** [comment]
Yeah, that's definitely still missing here. The main reason is that haven't quite gotten to that point with my app yet 🙂. At first glance it might look like we can just take the queue and serialize every operation in it, but unfortunately that doesn't really work in practice, because when the app is started back up, the operation actually needs to be repeated by the original requester (ie. Apollo Client or whatever else you're using) to make sure that the chain of observables/ promises is set up correctly. If you're using a client like Apollo client that means that it has to repeat the operations, and thus it has to serialize them for offline storage.
I don't think Apollo Client currently has such functionality, but it shouldn't be too hard to build it yourself by creating a small wrapper around Apollo Client which remembers all the mutations that didn't complete when the app was turned off, and repeat them when the app comes back online.

So, in summary here's what I think should happen:
1. Whenever a mutation is made, it is added to a queue (which is serialized at periodic intervals)
2. Mutations that succeed are removed from the queue
3. When the app is closed (rather than just going offline), all mutations are treated as failed.
4. When the app is turned on again all mutations in the queue are repeated again.

PS: For my app I've started working on a store (Atlas) that should have all that functionality out of the box. If you're interested, I can let you know once I have a testable prototype.

**Comment by MrLoh** [comment]
Sounds very interesting. I'm not quite there yet either, but will start to look into that soon. Would definitely be interested in your solution, keep me posted.

**Comment by bennyhobart** [comment]
Happy to lend a hand when we come up with a solution, it's a core requirement for an app I'm building

**Comment by cesaraaron** [comment]
Any updates on this?

**Comment by geminiyellow** [comment]
hope it working.

**Comment by MrLoh** [comment]
I am firing my mutations from redux action side effects (also helps to keep the UI code cleaner as the whole mutation logic is nicely separated). It is super easy to persist and replay redux actions, so that's a stable workaround, but that of course requieres using redux. For queries the problem usually doesn't arise as they are fired anew on app start anyways.

**Comment by wtrocki** [comment]
Happy to add this. It will be nice to use `apollo-cache-persist` as in most of the practical cases Apollo Cache will be persisted using this package so adding it here will work.

**Comment by wtrocki** [comment]
Done in: https://github.com/helfer/apollo-link-queue/pull/14

**Comment by wtrocki** [comment]
>  If you're using a client like Apollo client that means that it has to repeat the operations, and thus it has to serialize them for offline storage. I don't think Apollo Client currently has such functionality, but it shouldn't be too hard to build it yourself by creating a small wrapper around Apollo Client which remembers all the mutations that didn't complete when the app was turned off, and repeat them when the app comes back online.

I agree that this is an big problem on application restarts, but data that is currently saved as operation has everything that is needed to restore and replay any client side queries (mutations to be precise).
It could be responsibility of the apollo-link-queue + some extra package to reply that on top of the Apollo client, however then queue is not going to help at all. 

h4. 

Ideally queue should be responsible for saving it's operations  when closed. 
So at any point of time users will be able to see what operations are still not applied.
When operations will be resumed (open) data should be removed. 

However queue should not be initialized at startup with the data from storage as this is not going to work due 
to missing event's/promise chain. I tried to see how this could be saved and it's not so easy so natural way to resolve it is just to replicate the changes ouside link (directly with the client).

How this could work

1. Queue saves the data that is in flight
2. Page is refreshed
3. External package (needs to be build) will accept apollo client and will be able to replicate the changes that could land again on the queue or be processed and sent to server (depending on open/closed state on restart).

I still believe that queue will have significant role in staging this changes.
Without supporting restarts usage of the Apollo-link-queue is fairly limited in terms of offline support/application restarts performed changes could be lost.

**Comment by wtrocki** [comment]
I have published a project that using a queue concept but brings that idea further by supporting restarts and other features. See: https://github.com/graphql-heroes/offix/
